libnmstate/varlink/__init__.py

```diff
@@ -0,0 +1,292 @@
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program. If not, see <https://www.gnu.org/licenses/>.
+#
+import os
+import libnmstate
+import argparse
+import logging
+import errno
+import io
+import json
+import libnmstate.error as libnmError
+
+
+try:
+    import varlink
+
+except ModuleNotFoundError:
+    raise libnmError.NmstateDependencyError("python3 varlink module not found")
+
+
+logger = logging.getLogger()
+logger.setLevel(logging.DEBUG)
+log_stringio = io.StringIO()
+handler = logging.StreamHandler(log_stringio)
+handler.setLevel(logging.DEBUG)
+handler.setFormatter(
+    logging.Formatter(
+        '{"time": "%(asctime)s", "name": "%(name)s", '
+        + '"level": "%(levelname)s", "message": "%(message)s"}',
+        datefmt="%Y-%m-%d %H:%M:%S",
+    )
+)
+logger.addHandler(handler)
+
+service = varlink.Service(
+    vendor="Red Hat",
+    product="Nmstate",
+    version=libnmstate.__version__,
+    url="https://www.nmstate.io",
+    interface_dir=os.path.dirname(__file__),
+    namespaced=False,
+)
+
+
+def clean_log():
+    """
+    Cleaning previous logs
+    """
+    log_stringio.seek(0)
+    log_stringio.truncate()
+
+
+def validate_method_arguments(user_args, method_args):
+    """
+    Returns dictionary with validated arguments values.
+    """
+    kwargs = {}
+    for arg in user_args.keys():
+        if arg not in method_args or user_args[arg] == {}:
+            raise varlink.InvalidParameter(arg)
+        if user_args[arg] is not None:
+            kwargs[arg] = user_args[arg]
+    return kwargs
+
+
+def get_logs():
+    """
+    Returns list of logs in dictionary.
+    """
+    log_list = list()
+    log_stringio.seek(0)
+    for log in log_stringio.getvalue().replace("x\00", "").split("\n")[:-1]:
+        log_list.append(json.loads(log))
+    clean_log()
+    log_stringio.truncate(0)
+    return log_list
+
+
+def get_varlink_server(address):
+    """
+    Returns varlink server object
+    Checks if the varlink address already in use.
+    """
+    server = varlink.ThreadingServer(
+        address, ServiceRequestHandler, bind_and_activate=False
+    )
+    server.allow_reuse_address = True
+    try:
+        server.server_bind()
+        server.server_activate()
+    except OSError as exception:
+        server.shutdown()
+        server.server_close()
+        if exception.errno == errno.EADDRINUSE:
+            server = varlink.ThreadingServer(
+                address, ServiceRequestHandler, bind_and_activate=True
+            )
+        else:
+            logging.error(exception.strerror)
+            raise exception
+    return server
+
+
+def run_server(address):
+    """
+    Runs the varlink server in the specified the address
+    uses unix socket address and tcp address
```

Comment:
The address should a file path. We should not support unix abstract socket or tcp address.

Reply:
The other socket types clue be useful for other solutions using varlink for example using SSH port forwarding.

Reply:
IMO, we can support only file path initially and extend it later if needed. It seems we have different opinions about unix abstract socket or tcp address methods.